examples/tensorflow_estimator_simple.py

```diff
@@ -53,68 +53,49 @@ def create_optimizer(trial):
+        def optimizer(): return tf.keras.optimizers.SGD(learning_rate=sgd_lr,
+                                                        momentum=sgd_momentum)
 
     return optimizer
 
 
-def model_fn(trial, features, labels, mode):
-    logits = create_network(trial, features)
-
-    predictions = {
-        "classes": tf.argmax(input=logits, axis=1),
-        "probabilities": tf.nn.softmax(logits, name="softmax_tensor")
-    }
+def create_classifier(trial):
+    # We optimize the numbers of layers and their units.
 
-    if mode == tf.estimator.ModeKeys.PREDICT:
-        return tf.estimator.EstimatorSpec(mode=mode, predictions=predictions)
+    n_layers = trial.suggest_int('n_layers', 1, 3)
+    hidden_units = []
+    for i in range(n_layers):
+        n_units = trial.suggest_int('n_units_l{}'.format(i), 1, 128)
+        hidden_units.append(n_units)
 
-    loss = tf.compat.v1.losses.sparse_softmax_cross_entropy(labels=labels, logits=logits)
+    optimizer = create_optimizer(trial)
 
-    if mode == tf.estimator.ModeKeys.TRAIN:
-        optimizer = create_optimizer(trial)
-        train_op = optimizer.minimize(loss, tf.compat.v1.train.get_or_create_global_step())
-        return tf.estimator.EstimatorSpec(mode=mode, loss=loss, train_op=train_op)
+    model_dir = '{}/{}'.format(MODEL_DIR, trial.number)
+    classifier = tf.estimator.DNNClassifier(
+        feature_columns=[tf.feature_column.numeric_column('x', shape=[28*28])],
+        hidden_units=hidden_units,
+        model_dir=model_dir,
+        n_classes=10,
+        optimizer=optimizer)
 
-    eval_metric_ops = {
-        "accuracy": tf.compat.v1.metrics.accuracy(labels=labels,
-                                                  predictions=predictions["classes"])
-    }
-    return tf.estimator.EstimatorSpec(mode=mode, loss=loss, eval_metric_ops=eval_metric_ops)
+    return classifier
 
 
 def objective(trial):
-    (train_data, train_labels), (eval_data, eval_labels) = tf.keras.datasets.mnist.load_data()
-
-    train_data = train_data / np.float32(255)
-    train_labels = train_labels.astype(np.int32)
-
-    eval_data = eval_data / np.float32(255)
-    eval_labels = eval_labels.astype(np.int32)
-
-    model_dir = "{}/{}".format(MODEL_DIR, trial.number)
-    mnist_classifier = tf.estimator.Estimator(
-        model_fn=lambda features, labels, mode: model_fn(trial, features, labels, mode),
-        model_dir=model_dir)
-
-    train_input_fn = tf.compat.v1.estimator.inputs.numpy_input_fn(
-        x={"x": train_data}, y=train_labels, batch_size=BATCH_SIZE, num_epochs=None, shuffle=True)
+    classifier = create_classifier(trial)
 
-    mnist_classifier.train(input_fn=train_input_fn, steps=TRAIN_STEPS)
+    classifier.train(input_fn=input_fn, steps=TRAIN_STEPS)
 
-    eval_input_fn = tf.compat.v1.estimator.inputs.numpy_input_fn(
-        x={"x": eval_data}, y=eval_labels, num_epochs=1, shuffle=False)
+    eval_results = classifier.evaluate(input_fn=input_fn, steps=1)
```

Comment:
This code uses the training accuracy as the objective value while the previous code uses test accuracy. Please use the test data to avoid overfitting to the training data.

In addition, I think this code use only `BATCH_SIZE` images in the dataset because `steps` is set to `1` (c.f., [DNNClassifier.evaluate](https://www.tensorflow.org/api_docs/python/tf/estimator/DNNClassifier#evaluate)).  How about removing the `steps` argument and use the whole dataset? In that case, we can restrict the number of images using [`take()`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#take) method of `tfds.Dataset` as can be seen in https://github.com/optuna/optuna/blob/master/examples/tensorflow_eager_simple.py#L105.

Reply:
Thanks! I fixed them.